pkg/yurthub/storage/disk/storage.go

```diff
@@ -457,51 +503,172 @@ func (ds *diskStorage) Recover(key string) error {
+		}
+	}
+	if err := ds.fsOperator.Rename(tmpPath, path); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ds *diskStorage) recoverDir(tmpPath string) error {
+	if ok, err := fs.IsDir(tmpPath); err != nil || !ok {
+		return fmt.Errorf("failed at tmp path %s, isDir: %v, error: %v", tmpPath, ok, err)
+	}
+
+	tmpKey := strings.TrimPrefix(tmpPath, ds.baseDir)
+	key := getKey(tmpKey)
+	path := filepath.Join(ds.baseDir, key)
+	if fs.IfExists(path) {
+		if ok, err := fs.IsDir(path); err != nil || !ok {
+			return fmt.Errorf("failed at origin path %s, isDir: %v, error: %v", path, ok, err)
+		}
+		if err := ds.fsOperator.DeleteDir(path); err != nil {
+			return fmt.Errorf("failed to delete dir at %s, %v", path, err)
+		}
+	}
+	if err := ds.fsOperator.Rename(tmpPath, path); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ds *diskStorage) lockKey(key storageKey) bool {
+	keyStr := key.Key()
 	ds.Lock()
 	defer ds.Unlock()
-	if _, ok := ds.keyPendingStatus[key]; ok {
-		klog.Infof("key(%s) storage is pending, just skip it", key)
+	if _, ok := ds.keyPendingStatus[keyStr]; ok {
+		klog.Infof("key(%s) storage is pending, just skip it", keyStr)
 		return false
 	}
 
 	for pendingKey := range ds.keyPendingStatus {
-		if len(key) > len(pendingKey) {
-			if strings.Contains(key, fmt.Sprintf("%s/", pendingKey)) {
-				klog.Infof("key(%s) storage is pending, skip to store key(%s)", pendingKey, key)
+		if len(keyStr) > len(pendingKey) {
+			if strings.Contains(keyStr, fmt.Sprintf("%s/", pendingKey)) {
+				klog.Infof("key(%s) storage is pending, skip to store key(%s)", pendingKey, keyStr)
 				return false
 			}
 		} else {
-			if strings.Contains(pendingKey, fmt.Sprintf("%s/", key)) {
-				klog.Infof("key(%s) storage is pending, skip to store key(%s)", pendingKey, key)
+			if strings.Contains(pendingKey, fmt.Sprintf("%s/", keyStr)) {
+				klog.Infof("key(%s) storage is pending, skip to store key(%s)", pendingKey, keyStr)
 				return false
 			}
 		}
 	}
-	ds.keyPendingStatus[key] = struct{}{}
+	ds.keyPendingStatus[keyStr] = struct{}{}
 	return true
 }
 
-func (ds *diskStorage) unLockKey(key string) {
+func (ds *diskStorage) ifFresherThan(oldObj []byte, newRV uint64) (bool, error) {
+	// check resource version
+	unstructuredObj := &unstructured.Unstructured{}
+	curObj, _, err := ds.serializer.Decode(oldObj, nil, unstructuredObj)
+	if err != nil {
+		return false, fmt.Errorf("failed to decode obj, %v", err)
+	}
+	curRv, err := ObjectResourceVersion(curObj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get rv of obj, %v", err)
+	}
+	if newRV < curRv {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (ds *diskStorage) unLockKey(key storageKey) {
 	ds.Lock()
 	defer ds.Unlock()
-	delete(ds.keyPendingStatus, key)
+	delete(ds.keyPendingStatus, key.Key())
+}
+
+func ifEnhancement(baseDir string, fsOperator fs.FileSystemOperator) (bool, error) {
+	resDirs, err := fsOperator.List(baseDir, fs.ListModeDirs, false)
```

Comment:
The format of dir's name under `baseDir` is `<component>`, and doesn't include `<Resource.Version.Group>` info.

Reply:
fixed